drivers/mil_passive_sonar/mil_passive_sonar/passive_sonar_driver.py

```diff
@@ -0,0 +1,315 @@
+        except serial.SerialException as e:
+            print  e
+            return {}
+
+        time = rospy.Time.now()
+        signals = receive_signals(self.ser, self.receiver_count, self.signal_sz, 
+                                  self.scalar_sz, self.signal_bias)
+        heading = self.multilaterator.getPulseLocation(self.get_dtoa(signals))
+
+        # Add heaing observation to buffers if the signals are above the variance threshold
+        variance = np.var(signals)
+        if variance > self.min_variance:
+            p0, R = getHydrophonePose(time)
+            R = tf.transformations.quaternion_matrix(R)
+            map_offset = R.dot(heading)
+            p1 = p0 + map_offset
+
+            self.visualize_heading(p0, p1, bgra=[1.0, 0, 0, 0.50], length=2.0)
+
+            self.heading_start = np.append(self.heading_start, p0, axis=0)
+            self.heading_end = np.append(self.heading_end, p1, axis=0)
+            self.observation_variances = np.append(self.observation_variances, variance)
+
+            # delete softest samples if we have over max_observations
+            if len(self.line_array) >= self.observation_buffer_size:
+                softest_idx = np.argmin(self.observation_variances)
+                self.heading_start = np.delete(self.line_array, softest_idx, axis=0)
+                self.heading_end = np.delete(self.line_array, softest_idx, axis=0)
+                self.observation_variances = np.delete(self.observation_variances, softest_idx, axis=0)
+
+        header = make_header(frame=self.locating_frame)
+        heading = {'x' : heading[0], 'y' : heading[1],  'z' : heading[2]}
+        return {'header' : header, 'heading' : heading}
+
+    def estimate_pinger_position(self, req):
+        '''
+        Uses a buffer of prior observations to estimate the position of the pinger as the intersection
+        of a set of 3d lines in the least-squares sense.
+        '''
+        assert len(self.heading_start) > 1
+        self.pinger_postion = ls_line_intersection3d(self.heading_start, self.heading_end)
+        self.visualize_pinger_pos_estimate()
+        pinger_position = numpy_to_point(self.pinger_position)
+        return {'pinger_position' : pinger_position, 'num_samples' : len(self.heading_start)}
+
+    def set_freq(self, req):
+        '''
+        Sets the assumed frequency (in absence of noise) of the signals to received by the driver
+        '''
+        self.target_freq = req.frequency
+        self.heading_start = np.empty((0, 3), float)
+        self.heading_end = np.empty((0, 3), float)
+        self.sample_variances = np.empty((0, 1), float)
+        self.pinger_position = np.array([np.NaN, np.NaN, np.NaN])
+        return {}
+
+    def reset_position_estimate(self, req):
+        '''
+        Clears all the heading and amplitude buffers and makes the position estimate NaN
+        '''
+	self.heading_start = np.empty((0, 3), float)
+	self.heading_end = np.empty((0, 3), float)
```

Comment:
There's more spacing issues here. I know python will break under certain conditions if the spacing is off by too much.